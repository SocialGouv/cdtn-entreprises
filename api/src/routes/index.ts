import Router from "koa-router";
import { search, searchEntreprise, searchEtablissement } from "../lib";

export const router = new Router();

export const API_PREFIX = "/api/v1";

router.get(`${API_PREFIX}/search`, async (ctx) => {
  const { q: query, a: address, l: limit } = ctx.query;

  if (!query) {
    ctx.throw(400, `query parameter q is required`);
  }

  try {
    const entreprises = await search(
      query as string,
      address as string,
      parseInt(limit as string),
      true,
      true
    );
    ctx.body = { entreprises };
  } catch (err) {
    console.log(JSON.stringify(err));
    ctx.throw(500);
  }
});

const siretRE = new RegExp(/^\d{14}$/);
const validateSiretFormat = (siret: string) => siret.match(siretRE);

router.get(`${API_PREFIX}/etablissement/:siret`, async (ctx) => {
  const { siret } = ctx.params;

  if (!siret) {
    ctx.throw(400, "siret required");
  } else if (!validateSiretFormat(siret)) {
    ctx.throw(400, "invalid siret");
  }

  try {
    const etablissement = await searchEtablissement(siret as string);
    if (etablissement) {
      ctx.body = etablissement;
    } else {
      ctx.throw(404);
    }
  } catch (err) {
    if (err.status) {
      throw err;
    } else {
      console.log(JSON.stringify(err));
      ctx.throw(500);
    }
  }
});

const sirenRE = new RegExp(/^\d{9}$/);
const validateSirenFormat = (siren: string) => siren.match(sirenRE);

router.get(`${API_PREFIX}/entreprise/:siren`, async (ctx) => {
  const { siren } = ctx.params;

  if (!siren) {
    ctx.throw(400, "siren required");
  } else if (!validateSirenFormat(siren)) {
    ctx.throw(400, "invalid siren");
  }

  try {
    const entreprise = await searchEntreprise(siren as string);
    if (entreprise) {
      ctx.body = entreprise;
    } else {
      ctx.throw(404);
    }
  } catch (err) {
    if (err.status) {
      throw err;
    } else {
      console.log(JSON.stringify(err));
      ctx.throw(500);
    }
  }
});
