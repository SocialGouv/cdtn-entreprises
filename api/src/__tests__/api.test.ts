import supertest from "supertest";
import * as http from "http";
import Koa from "koa";
import { API_PREFIX, router } from "../routes";

export const app = new Koa();
app.use(router.routes());

const apptest = supertest(http.createServer(app.callback()));

describe("Test endpoints", () => {
  test("generic search", async (done) => {
    const { status, body } = await apptest.get(`${API_PREFIX}/search?q=zara`);

    expect(status).toBe(200);
    expect(body).toMatchSnapshot();

    done();
  });
});
