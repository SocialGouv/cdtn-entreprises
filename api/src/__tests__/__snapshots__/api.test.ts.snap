// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test entreprise search correct siren 1`] = `
Object {
  "activitePrincipale": "Fabrication d'articles en fils métalliques, de chaînes et de ressorts",
  "conventions": Array [
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635597",
      "idcc": 45,
      "mtime": 1562873918,
      "shortTitle": "Caoutchouc",
      "texte_de_base": "KALITEXT000005673838",
      "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
    },
    Object {
      "idcc": 9999,
      "shortTitle": "",
    },
  ],
  "etablissements": 36,
  "highlightLabel": "MANUF FRANC PNEUMATIQ MICHELIN",
  "label": "MANUF FRANC PNEUMATIQ MICHELIN",
  "matching": 26,
  "matchingEtablissement": Object {
    "address": "Rue du Xay 88190 Golbey",
    "siret": "85520050700108",
  },
  "simpleLabel": "MANUF FRANC PNEUMATIQ MICHELIN",
  "siren": "855200507",
}
`;

exports[`Test etablissement search correct siret 1`] = `
Object {
  "activitePrincipale": "Fabrication d'articles en fils métalliques, de chaînes et de ressorts",
  "address": "Rue du Xay 88190 Golbey",
  "convention": Object {
    "etat": "VIGUEUR_ETEN",
    "id": "KALICONT000005635597",
    "idcc": 45,
    "mtime": 1562873918,
    "shortTitle": "Caoutchouc",
    "texte_de_base": "KALITEXT000005673838",
    "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
    "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
  },
  "etablissements": 36,
  "highlightLabel": "MANUF FRANC PNEUMATIQ MICHELIN",
  "label": "MANUF FRANC PNEUMATIQ MICHELIN",
  "matching": 1,
  "simpleLabel": "MANUF FRANC PNEUMATIQ MICHELIN",
  "siren": "855200507",
  "siret": "85520050700108",
}
`;

exports[`Test search generic search 1`] = `
Object {
  "activitePrincipale": "Fabrication d'articles en fils métalliques, de chaînes et de ressorts",
  "conventions": Array [
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635597",
      "idcc": 45,
      "mtime": 1562873918,
      "shortTitle": "Caoutchouc",
      "texte_de_base": "KALITEXT000005673838",
      "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
    },
  ],
  "etablissements": 36,
  "highlightLabel": "MANUF FRANC PNEUMATIQ <b><u>MICHELIN</b></u>",
  "label": "MANUF FRANC PNEUMATIQ MICHELIN",
  "matching": 14,
  "matchingEtablissement": Object {
    "address": "Rue du Xay 88190 Golbey",
    "siret": "85520050700108",
  },
  "simpleLabel": "MANUF FRANC PNEUMATIQ MICHELIN",
  "siren": "855200507",
}
`;
